src/routes/styles.ts

```diff
@@ -0,0 +1,156 @@
+import createError from 'fastify-error'
+import got from 'got'
+
+import { normalizeStyleURL } from '../lib/mapbox_urls'
+import {
+  OfflineStyle,
+  StyleJSON,
+  createIdFromStyleUrl,
+  validate,
+} from '../lib/stylejson'
+
+interface GetStylesQuerystring {
+  limit?: number
+}
+
+interface GetStyleParams {
+  styleId: string
+}
+
+interface PutStyleParams {
+  styleId: string
+}
+
+type PutStyleBody = OfflineStyle | StyleJSON
+
+interface DeleteStyleParams {
+  styleId: string
+}
+
+type PostStyleBody = { accessToken?: string } & (
+  | { url: string }
+  | { filepath: string }
+  | { style: StyleJSON }
+)
+
+const InvalidStyleError = createError(
+  'FST_INVALID_STYLE',
+  'Invalid style: %s',
+  400
+)
+
+const InvalidRequestBodyError = createError(
+  'FST_INVALID_REQUEST_BODY',
+  'Invalid request body: %s',
+  400
+)
+
+function createInvalidStyleError(err: unknown) {
+  return new InvalidStyleError((err as Error).message)
+}
+
+function validateStyle(style: unknown): asserts style is StyleJSON {
+  try {
+    validate(style)
+  } catch (err) {
+    throw createInvalidStyleError(err)
+  }
+}
+
+const styles: FastifyPluginAsync = async function (fastify) {
+  fastify.get<{ Querystring: GetStylesQuerystring }>(
+    '/',
+    async function (request) {
+      return request.api.listStyles(request.query.limit)
+    }
+  )
+
+  fastify.post<{ Body: PostStyleBody }>('/', async function (request, reply) {
+    let style: unknown
+    let id: string | undefined
+    const { accessToken } = request.body
+
+    if ('filepath' in request.body) {
+      try {
+        const parsedStyle = JSON.parse(
+          await fs.readFile(request.body.filepath, 'utf-8')
+        )
+
+        if (!parsedStyle.id) {
+          throw new Error('Styles imported via file must have an id field')
+        }
+
+        id = parsedStyle.id
+      } catch (err) {
+        throw createInvalidStyleError(err)
+      }
+    } else if ('url' in request.body) {
+      try {
+        const { url } = request.body
+
+        // This will throw if the url is a mapbox style url and an access token is not provided
+        // Ideally prevented via client-side code but just in case
+        const upstreamUrl = normalizeStyleURL(url, accessToken)
+
+        style = (await got(upstreamUrl).json()) as any
+        id = createIdFromStyleUrl(url)
+      } catch (err) {
+        throw createInvalidStyleError(err)
+      }
+    } else if ('style' in request.body) {
+      style = request.body.style
+    } else {
+      throw new InvalidRequestBodyError(
+        'Body must have one of the following fields: style, filepath, url'
+      )
+    }
+
+    validateStyle(style)
+
+    // TODO: Should we catch the missing access token issue before calling this? i.e. check if `url` or any of `style.sources` are Mapbox urls
+    // `createStyle` will catch these but may save resources in the db before that occurs
+    const stylejson = await request.api.createStyle(style, { id, accessToken })
```

Comment:
Yeah I believe we have a db field for the Style table for this exact thing. I don't think i've actually written anything that handles this at the moment though.

> We haven't discussed when that update should happen

yeah i think we briefly discussed this but would be good to think through. I like the idea of it not being automatic but rather via an explicit update action from the user's end, but i don't have the best context of what the user would expect from this kind of functionality.

Reply:
> Yeah I believe we have a db field for the Style table for this exact thing.

Correction: we don't have a field for this 😅 guessing it would make sense to add though

Reply:
Partially addressed via https://github.com/digidem/mapeo-mapserver/commit/3cdb1c9b2df239248ab91c9038a1ea903f895a94